File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
n = int(input())
graph = []
num = []

for i in range(n):
    graph.append(list(map(int, input())))

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]


def dfs(x, y):
    if x < 0 or x >= n or y < 0 or y >= n:
        return 0

    if graph[x][y] == 1:
        global count
        count += 1
        graph[x][y] = 0
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            dfs(nx, ny)
        return 1
    return 0


count = 0
result = 0

for i in range(n):
    for j in range(n):
        if dfs(i, j) == 1:
            num.append(count)
            result += 1
            count = 0

num.sort()
print(result)
for i in range(len(num)):
    print(num[i])

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void DFS(int x, int y, vector<vector<int>> &graph, vector<vector<bool>> &visited, int &cnt, int N) {
    int dx[4] = {-1, 0, 1, 0};
    int dy[4] = {0, -1, 0, 1};

    visited[x][y] = true;
    cnt += 1;

    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];

        if (nx >= 0 && nx < N && ny >= 0 && ny < N) {
            if (!visited[nx][ny] && graph[nx][ny] == 1) {
                DFS(nx, ny, graph, visited, cnt, N);
            }
        }
    }
}

int main() {
    int N;
    cin >> N;

    vector<vector<int>> graph(N, vector<int>(N));
    vector<vector<bool>> visited(N, vector<bool>(N, false));
    vector<int> result;

    for (int i = 0; i < N; i++) {
        string row;
        cin >> row;
        for (int j = 0; j < N; j++) {
            graph[i][j] = row[j] - '0';
        }
    }

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (!visited[i][j] && graph[i][j] == 1) {
                int cnt = 0;
                DFS(i, j, graph, visited, cnt, N);
                result.push_back(cnt);
            }
        }
    }

    sort(result.begin(), result.end());

    cout << result.size() << '\n';
    for (int num : result) {
        cout << num << '\n';
    }

    return 0;
}
